libr/anal/p/anal_avr.c

```diff
@@ -17,35 +17,277 @@
+		"$b3,hf,=,"			// half carry flag
+		"$o,vf,=,"			// overflow flag
+		"vf,nf,^,sf,=",			// sign flag : xor(vf, nf)
+		r, d, r, d);
+}
+
+// ALIAS: Opcode (clr) when Rra and Rrb are the same
+INST_HANDLER (eor) {
+	int d = ((buf[0] & 0xf0) >> 4) | ((buf[1] & 1) << 4);
+	int r = (buf[0] & 0xf) | ((buf[1] & 2) << 3);
+
+	r_strbuf_setf (&op->esil, "r%d,r%d,^=,$z,zf,=,r%d,0x80,&,!,!,nf,=,nf,sf,=,0,vf,=", r, d, d);
+}
+
+INST_HANDLER (movw) {
+	int d = (buf[0] & 0xf0) >> 3;
+	int r = (buf[0] & 0x0f) << 1;
+
+	r_strbuf_setf (&op->esil, "r%d,r%d,=,r%d,r%d,=", r, d, r + 1, d + 1);
+}
+
+INST_HANDLER (nop) {
+}
+
+INST_HANDLER (out) {
+	op->type2 = 1;
+	op->val = (buf[0] & 0x0f) | (((buf[1] >> 1) & 0x03) << 4);
+	
+	int r = ((buf[0] >> 4) & 0x0f) | ((buf[1] & 0x01) << 4);
+
+	// launch esil trap (communicate upper layers about this I/O)
+	switch (op->val) {
+	case 0x3f: /* SREG */ r_strbuf_setf (&op->esil, "r%d,sreg,=", r); break;
+	case 0x3e: /* SPH  */ r_strbuf_setf (&op->esil, "r%d,sph,=",  r); break;
+	case 0x3d: /* SPL  */ r_strbuf_setf (&op->esil, "r%d,spl,=",  r); break;
+	default:
+		r_strbuf_setf (&op->esil, "2,$");
+	}
+}
+
+INST_HANDLER (rcall) {
+	register int32_t offset;
+
+	// target offset
+	offset = ((((buf[1] & 0xf) << 8) | buf[0]) + 1) << 1;
+	if (offset & 0x100)
+		offset |= 0xfffff000;
+	op->jump = op->addr + offset;
+	op->fail = 0;
+
+	// cycles!
+	if (strncasecmp (anal->cpu, "ATtiny", 6)) {
+		op->cycles = 4;	// ATtiny is always slow
+	} else {
+		// PC size decides required runtime!
+		op->cycles = cpu->pc_bits <= 16 ? 3 : 4;
+		if (strncasecmp (anal->cpu, "ATxmega", 7))
+			op->cycles--;	// ATxmega optimizes one cycle
+	}
+
+	r_strbuf_setf (
+		&op->esil,
+		"pc,"			// esil is already pointing to the
+					// next instruction (@ret)
+		"sp,-%d,+,"		//   and dec by (PC_SIZE-1) SP
+		"_sram,+,"              //   and point to the SRAM!
+		"=[%d],"		// store ret@ in stack
+		"sp,-%d,+,"		// decrement stack pointer
+		"sp,=,"			// store SP
+		"%d,pc,=",		// jump!
+		cpu->pc_size - 1, cpu->pc_size,
+		cpu->pc_size, op->jump);
+}
+
+INST_HANDLER (ret) {
+	if (cpu->pc_size > 2)	// if we have a bus bigger than 16 bit
+		op->cycles++;	// (i.e. a 22-bit bus), add one extra cycle
+	op->eob = true;
+
+	r_strbuf_setf (
+		&op->esil,
+		"sp,"			// load stack pointer
+		"sp,1,+,"		//   and inc by 1 SP
+		"_sram,+,"              //   and point to the SRAM!
+		"[%d],"			// read ret@ from the stack
+		"pc,=,"			// update PC with [SP]
+		"sp,%d,+,"		// post increment stack pointer
+		"sp,=,",		// store incremented SP
+		cpu->pc_size, cpu->pc_size);
+}
+
+INST_HANDLER (reti) {
+	INST_CALL (ret);
+
+	//XXX: There are not privileged instructions in ATMEL/AVR
+	// op->family = R_ANAL_OP_FAMILY_PRIV;
+
+	// RETI: The I-bit is cleared by hardware after an interrupt
+	// has occurred, and is set by the RETI instruction to enable
+	// subsequent interrupts
+	r_strbuf_append (&op->esil, ",1,if,=");
+}
+
+INST_HANDLER (rjmp) {
+	register int32_t offset;
```

Comment:
dont use compiler hints like register. they are mostly ignored in modern compiler optimizations